File: turing/binary_tree/sumOfLeftLeaves.js
/**
 * Definition for a binary tree node:
 * function TreeNode(val, left, right) {
 *   this.val = (val===undefined ? 0 : val)
 *   this.left = (left===undefined ? null : left)
 *   this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
class TreeNode {
  constructor(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
  }
}

// function sumOfLeftLeaves(root) {
//   if (!root) return 0;
//   let sum = 0;
//   if (root.left && !root.left.left && !root.left.right) {
//     sum += root.left.val;
//   }
//   sum += sumOfLeftLeaves(root.left) + sumOfLeftLeaves(root.right);
//   return sum;
// };

var sumOfLeftLeaves = function(root) {
  if (!root) return 0;
  if (root.left && !root.left.left && !root.left.right) return root.left.val + sumOfLeftLeaves(root.right);
  return sumOfLeftLeaves(root.left) + sumOfLeftLeaves(root.right);
};

//e.g. 1) [3,9,20,null,null,15,7], return 24
// diagram:
//     3
//    / \
//   9  20
//     /  \
//    15   7
x = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
console.log('Sum of left leaves: ' + sumOfLeftLeaves(x));

//e.g. 2) Binary tree: [5,3,6,2,4,null,null,null,1], return: 9
// diagram:
//       5
//     /   \
//    3     6
//   / \   / \
//  2   4 1   null
// /
// 1
x = new TreeNode(5, new TreeNode(3, new TreeNode(2), new TreeNode(4, new TreeNode(1), null)), new TreeNode(6)); // return 9
console.log('Sum of left leaves: ' + sumOfLeftLeaves(x));

File: turing/binary_tree/sumOfLeftLeaves.py
# Definition for a binary tree node.
class TreeNode(object):
  def __init__(self, val):
    self.value = val
    self.left = None
    self.right = None

class Solution(object):
  def sumOfLeftLeaves(self, root):
    """
    :type root: TreeNode
    :rtype: int
    """
    if root is None:
      return 0
    sum = 0
    if root.left and not root.left.left and not root.left.right:
      sum += root.left.value
    sum += self.sumOfLeftLeaves(root.left) + self.sumOfLeftLeaves(root.right)
    return sum

#Input: root = [3,9,20,null,null,15,7]
#Output: 24
#Explanation: There are two left leaves in the binary tree, with values 9 and 15 respectively. Return 24.
#diagram:
#     3
#    / \
#   9  20
#     /  \
#    15   7
x = TreeNode(3)
x.left = TreeNode(9)
x.right = TreeNode(20)
x.right.left = TreeNode(15)
x.right.right = TreeNode(7)
print(Solution().sumOfLeftLeaves(x))

#Input: root = [5,3,6,2,4,null,null,1]
#Output: 6
#Explanation: There are three left leaves in the binary tree, with values 3, 2 and 1 respectively. Return 6.
#diagram:
#        5
#       / \
#      3   6
#     / \
#    2   4
#       /
#      1
x = TreeNode(5)
x.left = TreeNode(3)
x.right = TreeNode(6)
x.left.left = TreeNode(2)
x.left.right = TreeNode(4)
x.left.right.left = TreeNode(1)
print(Solution().sumOfLeftLeaves(x))
